app/main.py

```diff
@@ -1,3 +1,31 @@
-def cache(func):
-    # Write your code here
-    pass
+from typing import Callable
+
+
+def cache(func: Callable) -> Callable:
+    previous_runs = {}
+
+    def inner(*args) -> object:
+        nonlocal previous_runs
+        hash_key = hash_function(func, args)
+        if hash_key in previous_runs.keys():
+            print("Getting from cache")
+            return previous_runs[hash_key]
+        else:
+            print("Calculating new result")
+            result = func(*args)
+            previous_runs[hash_key] = result
+        return result
+
+    return inner
+
+
+def hash_function(function_to_hash: Callable, args: tuple) -> str:
```

Comment:
> External function was created to develop it in advance, if hashing will be required to hide a name of function or arguments.

Good point, but we don't need it here. Unfortunately, in real work cases it might not be approved for the task description, so do I, sorry :(

Reply:
Fixed in next commit